server/cmwell-ws/app/security/AuthCache.scala

```diff
@@ -26,23 +26,40 @@ import play.api.libs.json.{JsValue, Json}
 
 import scala.concurrent.duration._
 import scala.concurrent.{Await, ExecutionContext, Future}
+import scala.util.Try
 
 @Singleton
 class AuthCache @Inject()(crudServiceFS: CRUDServiceFS)(implicit ec: ExecutionContext) extends LazyLogging {
   // TODO Do not Await.result... These should return Future[Option[JsValue]]]
-  def getRole(roleName: String, nbg: Boolean): Option[JsValue] = Await.result(data(nbg).getAndUpdateIfNeeded.map(_.roles.get(roleName)).recoverWith { case _ =>
-    logger.warn(s"AuthCache Graceful Degradation: Search failed! Trying direct read for Role($roleName):")
-    getFromCrudAndExtractJson(s"/meta/auth/roles/$roleName", nbg)
-  }, 5.seconds)
+  def getRole(roleName: String, nbg: Boolean): Option[JsValue] = {
+    val esDependingFut = data(nbg).getAndUpdateIfNeeded.map(_.roles.get(roleName))
+    Await.result(
+      cmwell.util.concurrent.timeoutOptionFuture(esDependingFut, 3.seconds).flatMap {
+        case Some(roleInfotonOpt) => Future.successful(roleInfotonOpt)
+        case None =>
+          logger.warn(s"AuthCache Graceful Degradation: Search failed! Trying direct read for Role($roleName):")
+          getFromCasAndExtractJson(s"/meta/auth/roles/$roleName", nbg)
+      }
+      , 6.seconds
+    )
+  }
 
-  def getUserInfoton(userName: String, nbg: Boolean): Option[JsValue] = Await.result(data(nbg).getAndUpdateIfNeeded.map(_.users.get(userName)).recoverWith { case _ =>
-    logger.warn(s"AuthCache Graceful Degradation: Search failed! Trying direct read for User($userName):")
-    getFromCrudAndExtractJson(s"/meta/auth/users/$userName", nbg)
-  }, 5.seconds)
+  def getUserInfoton(userName: String, nbg: Boolean): Option[JsValue] = {
```

Comment:
can we factor out commonalities of `getRole` and `getUserInfoton`?
code is almost identical.

Reply:
I'll give it a shot.